### PR TITLE
setup: facts: linux: respect partitions if device is nvme, cciss or loop.

### PR DESCRIPTION
**Pull request by k0ste**
##### SUMMARY

Current match don't work with "non-standard" devices like NVMe.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

setup

##### ANSIBLE VERSION

```
ansible 2.3.2.0
  config file = /home/k0ste/.ansible.cfg
  configured module search path = [u'/home/k0ste/ansible/my_modules', u'/home/k0ste/ceph-ansible/library']
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]
```

##### ADDITIONAL INFORMATION

Before:

```json
        "ansible_devices": {
            "nvme0n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Lite-On Technology Corporation M8Pe Series NVMe SSD (rev 01)", 
                "model": "PLEXTOR PX-128M8PeY", 
                "partitions": {}, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "250069680", 
                "sectorsize": "512", 
                "size": "119.24 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
            "nvme1n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Intel Corporation PCIe Data Center SSD (rev 01)", 
                "model": "INTEL SSDPEDMD400G4", 
                "partitions": {}, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "781422768", 
                "sectorsize": "512", 
                "size": "372.61 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
            "nvme2n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Lite-On Technology Corporation M8Pe Series NVMe SSD (rev 01)", 
                "model": "PLEXTOR PX-1TM8PeY", 
                "partitions": {}, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "2000409264", 
                "sectorsize": "512", 
                "size": "953.87 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
```

After:

```json
        "ansible_devices": {
            "nvme0n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Lite-On Technology Corporation M8Pe Series NVMe SSD (rev 01)", 
                "model": "PLEXTOR PX-128M8PeY", 
                "partitions": {
                    "nvme0n1p1": {
                        "holders": [], 
                        "sectors": "58720256", 
                        "sectorsize": 512, 
                        "size": "28.00 GB", 
                        "start": "2048", 
                        "uuid": null
                    }, 
                    "nvme0n1p2": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "58722304", 
                        "uuid": null
                    }, 
                    "nvme0n1p3": {
                        "holders": [], 
                        "sectors": "58720256", 
                        "sectorsize": 512, 
                        "size": "28.00 GB", 
                        "start": "60819456", 
                        "uuid": null
                    }, 
                    "nvme0n1p4": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "119539712", 
                        "uuid": null
                    }, 
                    "nvme0n1p5": {
                        "holders": [], 
                        "sectors": "58720256", 
                        "sectorsize": 512, 
                        "size": "28.00 GB", 
                        "start": "121636864", 
                        "uuid": null
                    }, 
                    "nvme0n1p6": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "180357120", 
                        "uuid": null
                    }, 
                    "nvme0n1p7": {
                        "holders": [], 
                        "sectors": "58720256", 
                        "sectorsize": 512, 
                        "size": "28.00 GB", 
                        "start": "182454272", 
                        "uuid": null
                    }, 
                    "nvme0n1p8": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "241174528", 
                        "uuid": null
                    }
                }, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "250069680", 
                "sectorsize": "512", 
                "size": "119.24 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
            "nvme1n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Intel Corporation PCIe Data Center SSD (rev 01)", 
                "model": "INTEL SSDPEDMD400G4", 
                "partitions": {
                    "nvme1n1p1": {
                        "holders": [], 
                        "sectors": "62914560", 
                        "sectorsize": 512, 
                        "size": "30.00 GB", 
                        "start": "2048", 
                        "uuid": null
                    }, 
                    "nvme1n1p2": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "62916608", 
                        "uuid": null
                    }, 
                    "nvme1n1p3": {
                        "holders": [], 
                        "sectors": "62914560", 
                        "sectorsize": 512, 
                        "size": "30.00 GB", 
                        "start": "65013760", 
                        "uuid": null
                    }, 
                    "nvme1n1p4": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "127928320", 
                        "uuid": null
                    }, 
                    "nvme1n1p5": {
                        "holders": [], 
                        "sectors": "62914560", 
                        "sectorsize": 512, 
                        "size": "30.00 GB", 
                        "start": "130025472", 
                        "uuid": null
                    }, 
                    "nvme1n1p6": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "192940032", 
                        "uuid": null
                    }, 
                    "nvme1n1p7": {
                        "holders": [], 
                        "sectors": "62914560", 
                        "sectorsize": 512, 
                        "size": "30.00 GB", 
                        "start": "195037184", 
                        "uuid": null
                    }, 
                    "nvme1n1p8": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "257951744", 
                        "uuid": null
                    }
                }, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "781422768", 
                "sectorsize": "512", 
                "size": "372.61 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
            "nvme2n1": {
                "holders": [], 
                "host": "Non-Volatile memory controller: Lite-On Technology Corporation M8Pe Series NVMe SSD (rev 01)", 
                "model": "PLEXTOR PX-1TM8PeY", 
                "partitions": {
                    "nvme2n1p1": {
                        "holders": [], 
                        "sectors": "204800", 
                        "sectorsize": 512, 
                        "size": "100.00 MB", 
                        "start": "2048", 
                        "uuid": "aaaf3aa9-feab-4a6d-ada1-466f15e53494"
                    }, 
                    "nvme2n1p2": {
                        "holders": [], 
                        "sectors": "1935190671", 
                        "sectorsize": 512, 
                        "size": "922.77 GB", 
                        "start": "65218560", 
                        "uuid": null
                    }, 
                    "nvme2n1p3": {
                        "holders": [], 
                        "sectors": "62914560", 
                        "sectorsize": 512, 
                        "size": "30.00 GB", 
                        "start": "206848", 
                        "uuid": null
                    }, 
                    "nvme2n1p4": {
                        "holders": [], 
                        "sectors": "2097152", 
                        "sectorsize": 512, 
                        "size": "1.00 GB", 
                        "start": "63121408", 
                        "uuid": null
                    }
                }, 
                "removable": "0", 
                "rotational": "0", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "", 
                "sectors": "2000409264", 
                "sectorsize": "512", 
                "size": "953.87 GB", 
                "support_discard": "512", 
                "vendor": null
            }, 
```

### CODE REVIEW

**Comment by mkrizek** [comment]
cc @alikins

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 1]
```diff
@@ -585,7 +585,10 @@ def get_device_facts(self):
 
             d['partitions'] = {}
             for folder in os.listdir(sysdir):
-                m = re.search("(" + diskname + r"\d+)", folder)
+                if "nvme" or "cciss" or "loop" in diskname:
```
Comment (mscherer) [inline lib/ansible/module_utils/facts/hardware/linux.py]: Wouldn't it be more correct to verify if diskname start by those strings ?

Comment (k0ste) [reply lib/ansible/module_utils/facts/hardware/linux.py]: May be. This is "backport" from https://github.com/ceph/ceph-ansible/pull/2219. Regex was replaced by a match in string.

**Comment by alikins** [comment]
@k0ste could you provide some samples of what /sys/block/*/  looks like for nvme, cciss devices?

A tar of /sys/block/nvme2n1/*  or similar for reference and/or tests?

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 1 above)
Comment (alikins) [reply lib/ansible/module_utils/facts/hardware/linux.py]: I agree, startswith is less likely to collide with other device names.

**Review by alikins — CHANGES_REQUESTED** [review]

**Comment by k0ste** [comment]
@alikins, if I understood you correctly.

```shell
# for e in *; do echo -e "${e}: "; cat ${e}; done                                                                             
alignment_offset: 
0
badblocks: 

bdi: 
cat: bdi: Is a directory
capability: 
50
dev: 
259:9
device: 
cat: device: Is a directory
discard_alignment: 
512
eui: 
00 23 03 56 30 01 1f 99
ext_range: 
256
holders: 
cat: holders: Is a directory
inflight: 
       0        0
mq: 
cat: mq: Is a directory
nsid: 
1
nvme2n1p1: 
cat: nvme2n1p1: Is a directory
nvme2n1p2: 
cat: nvme2n1p2: Is a directory
power: 
cat: power: Is a directory
queue: 
cat: queue: Is a directory
range: 
0
removable: 
0
ro: 
0
size: 
2000409264
slaves: 
cat: slaves: Is a directory
stat: 
21969113     3595 4900362077  8275406 175699974   262567 18921026541 919523932        0 32739300 927841961
subsystem: 
cat: subsystem: Is a directory
trace: 
cat: trace: Is a directory
uevent: 
MAJOR=259
MINOR=9
DEVNAME=nvme2n1
DEVTYPE=disk
wwid: 
eui.0023035630011f99
```

**Comment by k0ste** [comment]
@alikins ready_for_review.

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 2]
```diff
@@ -615,7 +615,10 @@ def get_device_facts(self):
 
             d['partitions'] = {}
             for folder in os.listdir(sysdir):
-                m = re.search("(" + diskname + r"\d+)", folder)
+                if diskname.startswith(('nvme', 'cciss', 'loop')):
```
Comment (alikins) [inline lib/ansible/module_utils/facts/hardware/linux.py:618]: Needs tests.

**Review by alikins — CHANGES_REQUESTED** [review]
Needs test.

**Comment by mattclay** [comment]
CI failure in unit tests:

```
>       self.assertEqual(len(mtab_entries), 38)
E       AssertionError: 39 != 38

test/units/module_utils/facts/hardware/test_linux.py:97: AssertionError
```

**Comment by k0ste** [comment]
@mattclay, thanks.

**Comment by k0ste** [comment]
Now, as I can see CI failed by:

```
2018-02-21 06:04:30 ERROR: Command "scp -i test/cache/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -P 22 -q -r test/runner/setup/remote.sh ec2-user@ec2-184-72-179-85.compute-1.amazonaws.com:/tmp" returned exit status 1.
```

And all other tests passed.

**Comment by k0ste** [comment]
@alikins ready_for_review

**Comment by k0ste** [comment]
#39730

**Pull request closed without merging**